crates/pallet-domains/src/staking.rs

```diff
@@ -1203,20 +1203,44 @@ pub(crate) fn do_unlock_operator<T: Config>(operator_id: OperatorId) -> Result<u
 pub(crate) fn do_reward_operators<T: Config>(
     domain_id: DomainId,
     operators: IntoIter<OperatorId>,
-    mut rewards: BalanceOf<T>,
+    rewards: BalanceOf<T>,
 ) -> Result<(), Error> {
     DomainStakingSummary::<T>::mutate(domain_id, |maybe_stake_summary| {
         let stake_summary = maybe_stake_summary
             .as_mut()
             .ok_or(Error::DomainNotInitialized)?;
 
-        let distribution = Perbill::from_rational(One::one(), operators.len() as u32);
-        let reward_per_operator = distribution.mul_floor(rewards);
-        for operator_id in operators {
+        let total_count = operators.len() as u64;
+        // calculate the operator weights based on the number of times they are repeated in the original list.
+        let operator_weights = operators.into_iter().fold(
+            BTreeMap::<OperatorId, u64>::new(),
+            |mut acc, operator_id| {
+                let total_weight = match acc.get(&operator_id) {
+                    None => 1,
+                    Some(weight) => weight + 1,
+                };
+                acc.insert(operator_id, total_weight);
+                acc
+            },
+        );
+
+        let mut allocated_rewards = BalanceOf::<T>::zero();
+        let mut weight_balance_cache = BTreeMap::<u64, BalanceOf<T>>::new();
```

Comment:
`weight_balance_cache` seems unnecessary since `operator_weights` is `BtreeMap` and is de-duplicated

Reply:
It is to avoid rational calculations if we already have the operator_reward for a given weight rather than calculating the same over and over again

Reply:
Oh, indeed, I overlooked this...